text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+    situations as discussed in the Alternatives section, and that further adjustments needed to be made.
+
+And thus in early 2023 I began work on a 3rd revision of the API, a revision I am increasingly confident in for 2
+reasons:
+
+1.  It is nearly a direct translation of the `Allocator` API, which has been used within the Rust standard library for
+    years now.
+2.  A core trait providing functionality and a set of marker traits providing guarantees is much easier to deal with
+    than multiple traits each providing related but subtly different functionalities and guarantees.
+
+The ability to develop 3rd-party extensions for increased safety also confirms, to me, that @CAD97 was on the right
+track when they removed the strong typing, and on the wrong track when they attempted to bake in borrow-checking: if
+it's easy enough to add safety, then it seems better for the "core" API to be minimalist instead.
+
+
+#   Unresolved Questions
+
+##  (Major) How to make `StoreBox<T, S>` coercible?
+
+Unfortunately, at the moment, `Box` is only coercible because it stores a `NonNull<T>`, which is coercible. Splitting
+`NonNull<T>` into `S::Handle` and `<T as Pointee>::Metadata`, as `StoreBox<T, S>` does, leads to losing coercibility.
+
+A separate solution is required to regain coercibility, which is out of scope of this RFC, but would have to be solved
+if `StoreBox<T, S>` were to become `Box`, which seems preferable to keeping it separate.
+
+A suggestion would be to have a `TypedMetadata<T>` lang item, which would implement `CoerceUnsized` appropriately, and
+[the companion repository showcases](https://github.com/matthieu-m/storage/blob/main/src/extension/typed_metadata.rs)
+how building upon this `StoreBox` gains the ability to be `CoerceUnsized`. This is a topic for another RFC, however.
+
+
+##  (Medium) To `Clone`, to `Default`?
+
+The _Safety_ section of the [`Allocator`](https://doc.rust-lang.org/nightly/std/alloc/trait.Allocator.html#safety)
+documentation notes that a `Clone` of an `Allocator` must be interchangeable with the original, and that all allocated
+pointers must remain until the last of the clones or copies of the allocator is dropped.
+
+The standard library then proceed to require `A: Allocator + Clone` to clone a `Box` or a `Vec`, when arguably it is not
+necessary to have an interchangeable allocator, and instead _semantically_ an independent allocator is required.
+
+This RFC, instead, favors using the `Default` bound for the `Clone` implementation of `StoreBox`:
+
+1.  It matches the desired semantics better -- a brand new store is required, not an interchangeable one.
+2.  A cloneable `InlineStore` cannot match the semantics of `Clone` required of Allocators.
+
+`Default` does have the issue that it may not mesh well with `dyn Store` or `dyn Allocator`, and while `Clone` can
+reasonably be implemented for `&dyn Store`, or `Rc<dyn Store>`, such is not the case for `Default`.
```

Comment:
I still think that using `Clone` is a bit of a perversion in this case, as it violates the idea of clones being "equal".

I'd rather enumerate other alternatives before settling on it.

Reply:
FWIW, the `Clone` documentation doesn't say anything about expecting clones to be "equal," it just says it's for duplicating an object. And I see duplicating some storage as duplicating the storage even if handles from the original aren't valid in the clone. A "fully proper" clone which clones not just the storage but also the objects within (rather than just duplicating their bytes) is fundamentally not possible, so this seems like the reasonable interpretation of cloning the storage.

Reply:
I agree the documentation doesn't, it doesn't mean that people don't. When you need to document something because it clashes with people's preconceived ideas... perhaps you should step back and pick a different name.

This is why I prefer `Default` here: it's clear that a new independent instance is created. Unfortunately, it doesn't work well with references and the like :/